src/common/health.rs

```diff
@@ -50,55 +81,65 @@ impl Ready {
+                    continue;
+                };
+
+                shards.push((Shard::new(collection, shard), is_ready(state)));
             }
         }
 
-        let mut initial_unhealthy_shards = self.initial_unhelathy_shards.write().await;
-
-        match initial_unhealthy_shards.deref_mut() {
-            None => {
-                let unhealthy_shards =
-                    shards.into_iter().filter_map(|(collection, shard, state)| {
-                        if !matches!(state, ReplicaState::Active | ReplicaState::Listener) {
-                            Some(Replica::new(collection, shard))
-                        } else {
-                            None
-                        }
-                    });
-
-                initial_unhealthy_shards
-                    .insert(unhealthy_shards.collect())
-                    .is_empty()
-            }
+        shards
+    }
 
-            Some(initial_unhealthy_shards) => {
-                for (collection, shard, state) in shards {
-                    let replica = Replica::new(collection, shard);
+    async fn initialize_unhealthy(&self, shards: impl IntoIterator<Item = Shard>) -> bool {
+        let mut unhealthy_shards = self.unhealthy_shards.write().await;
 
-                    if matches!(state, ReplicaState::Active | ReplicaState::Listener) {
-                        initial_unhealthy_shards.remove(&replica);
-                    } else if is_first_run {
-                        initial_unhealthy_shards.insert(replica);
-                    }
-                }
+        unhealthy_shards.extend(shards);
+        self.is_initialized.store(true, atomic::Ordering::Relaxed);
 
-                initial_unhealthy_shards.is_empty()
-            }
+        unhealthy_shards.is_empty()
+    }
+
+    async fn remove_healthy(&self, shards: impl IntoIterator<Item = Shard>) -> bool {
+        let mut unhealthy_shards = self.unhealthy_shards.write().await;
+
+        for shard in shards {
+            unhealthy_shards.remove(&shard);
         }
+
+        unhealthy_shards.is_empty()
     }
 }
 
 #[derive(Clone, Debug, Eq, PartialEq, Hash)]
-pub struct Replica {
+struct Shard {
     collection: CollectionId,
     shard: ShardId,
 }
 
-impl Replica {
-    pub fn new(collection: CollectionId, shard: ShardId) -> Self {
-        Self { collection, shard }
+impl Shard {
+    pub fn new(collection: impl Into<CollectionId>, shard: ShardId) -> Self {
+        Self {
+            collection: collection.into(),
+            shard,
+        }
     }
 }
+
+fn is_ready(state: ReplicaState) -> bool {
+    matches!(state, ReplicaState::Active | ReplicaState::Listener) // TODO: `ReplicaState::Initializing`?
+}
+
+fn filter(
+    shards: impl IntoIterator<Item = (Shard, bool)>,
+    filter: bool,
+) -> impl Iterator<Item = Shard> {
+    shards.into_iter().filter_map(move |(shard, is_ready)| {
+        if is_ready == filter {
+            Some(shard)
+        } else {
+            None
+        }
```

Comment:
There's [`then_some`](https://doc.rust-lang.org/std/primitive.bool.html#method.then_some) for this:

```rust
(is_ready == filter).then_some(shard)
```